motionplan/motionPlanner.go

```diff
@@ -99,180 +102,193 @@ func PlanFrameMotion(ctx context.Context,
 		&ConstraintInput{
 			StartInput: seedInputs,
 			EndInput:   target,
-			Frame:      mp.Frame(),
+			Frame:      mp.f,
 		},
-		planOpts.Resolution,
+		mp.planOpts.Resolution,
 	)
 	return ok
 }
 
-// node interface is used to wrap a configuration for planning purposes.
-type node interface {
-	// return the configuration associated with the node
-	Q() []frame.Input
-}
-
-type planReturn interface {
-	// return the steps in Input form
-	toInputs() [][]frame.Input
-	err() error
-}
-
-type basicNode struct {
-	q []frame.Input
-}
-
-func (n *basicNode) Q() []frame.Input {
-	return n.q
-}
-
-type costNode struct {
-	node
-	cost float64
+func (mp *planner) opt() *plannerOptions {
+	return mp.planOpts
 }
 
-func newCostNode(q []frame.Input, cost float64) *costNode {
-	return &costNode{&basicNode{q: q}, cost}
-}
-
-// nodePair groups together nodes in a tuple
-// TODO(rb): in the future we might think about making this into a list of nodes.
-type nodePair struct{ a, b node }
-
-func (np *nodePair) sumCosts() float64 {
-	a, aok := np.a.(*costNode)
-	if !aok {
-		return 0
+// smoothPath will try to naively smooth the path by picking points partway between waypoints and seeing if it can interpolate
+// directly between them. This will significantly improve paths from RRT*, as it will shortcut the randomly-selected configurations.
+// This will only ever improve paths (or leave them untouched), and runs very quickly.
+func (mp *planner) smoothPath(ctx context.Context, path []node) []node {
+	mp.logger.Debugf("running simple smoother on path of len %d", len(path))
+	if mp.planOpts == nil {
+		mp.logger.Debug("nil opts, cannot shortcut")
+		return path
 	}
-	b, bok := np.b.(*costNode)
-	if !bok {
-		return 0
+	if len(path) <= 2 {
+		mp.logger.Debug("path too short, cannot shortcut")
+		return path
 	}
-	return a.cost + b.cost
-}
 
-// EvaluatePlan assigns a numeric score to a plan that corresponds to the cumulative distance between input waypoints in the plan.
-func EvaluatePlan(plan planReturn, planOpts *plannerOptions) (totalCost float64) {
-	if errors.Is(plan.err(), errPlannerFailed) {
-		return math.Inf(1)
-	}
-	steps := plan.toInputs()
-	for i := 0; i < len(steps)-1; i++ {
-		_, cost := planOpts.DistanceFunc(&ConstraintInput{StartInput: steps[i], EndInput: steps[i+1]})
-		totalCost += cost
-	}
-	return totalCost
-}
+	// Randomly pick which quarter of motion to check from; this increases flexibility of smoothing.
+	waypoints := []float64{0.25, 0.5, 0.75}
 
-// GetSteps will determine the number of steps which should be used to get from the seed to the goal.
-// The returned value is guaranteed to be at least 1.
-// stepSize represents both the max mm movement per step, and max R4AA degrees per step.
-func GetSteps(seedPos, goalPos spatialmath.Pose, stepSize float64) int {
-	// use a default size of 1 if zero is passed in to avoid divide-by-zero
-	if stepSize == 0 {
-		stepSize = 1.
+	for i := 0; i < mp.planOpts.SmoothIter; i++ {
```

Comment:
Why are we smoothing this way? Is this something that you came across in a paper, or is it something that seems like an intuitive approach?  I'm not sure I'm sold on the random element of it.  It seems like you could just deterministically smooth it using some sort of binary search approach.  In any case, I think this is fine for now but might be something we want to flag for future work (along with sampling bias of some sort, be it informed-RRT or something else)

Reply:
Cool thats good to know.  I guess with this one we are bounded with our iterations anyway so it can't ever get that bad